src/dali/plugin/pair_converter/ccxt.py

```diff
@@ -0,0 +1,326 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from datetime import datetime, timedelta
+from time import sleep, time
+from typing import Any, Dict, List, NamedTuple, Optional, Union
+
+from ccxt import (binance, DDoSProtection, Exchange, ExchangeError, ExchangeNotAvailable, kraken, 
+    liquid, NetworkError, RequestTimeout)
+from rp2.logger import create_logger
+from rp2.rp2_decimal import RP2Decimal
+
+from dali.abstract_pair_converter_plugin import AbstractPairConverterPlugin
+from dali.historical_bar import HistoricalBar
+
+# Native format keywords
+_ID: str = "id"
+_BASE: str = "base"
+_QUOTE: str = "quote"
+
+# Time in ms
+_MINUTE: str = "1m"
+_FIVE_MINUTE: str = "5m"
+_FIFTEEN_MINUTE: str = "15m"
+_ONE_HOUR: str = "1h"
+_FOUR_HOUR: str = "4h"
+_ONE_DAY: str = "1d"
+_TIME_GRANULARITY: List[str] = [_MINUTE, _FIVE_MINUTE, _FIFTEEN_MINUTE, _ONE_HOUR, _FOUR_HOUR, _ONE_DAY]
+_TIME_GRANULARITY_IN_SECONDS: List[int] = [60, 300, 900, 3600, 14400, 86400]
+
+# Currently supported exchanges
+_BINANCE: str = "Binance.com"
+_KRAKEN: str = "Kraken"
+_LIQUID: str = "Liquid"
+_FIAT_EXCHANGE: str = "Exchangerate.host"
+_EXCHANGE_DICT: Dict[str, Any] = {_BINANCE: binance, _KRAKEN: kraken, _LIQUID: liquid}
+
+# Delay in fractional seconds before making a request to avoid too many request errors
+# Kraken states it has a limit of 1 call per second, but this doesn't seem to be correct.
+# It appears Kraken public API is limited to around 12 calls per minute.
+# There also appears to be a limit of how many calls per 2 hour time period.
+# Being authenticated lowers this limit.
+_REQUEST_DELAYDICT: Dict[str, float] = {_BINANCE: 0.0, _KRAKEN:5.1, _LIQUID:0}
+
+# Alternative Markets and exchanges for stablecoins or untradeable assets
+_ALTMARKET_EXCHANGES_DICT: Dict[str, str] = {"USDTUSD": _KRAKEN, "SOLOXRP": _LIQUID}
+_ALTMARKET_BY_BASE_DICT: Dict[str, str] = {"USDT": "USD", "SOLO": "XRP"}
+
+# Time constants
+_MS_IN_SECOND: int = 1000
+
+
+class StableCoin(NamedTuple):
+    symbol: str
+    fiat_code: str
+
+
+class AssetPairAndTimestamp(NamedTuple):
+    timestamp: datetime
+    from_asset: str
+    to_asset: str
+    exchange: str
+
+
+class AssetPairAndHistoricalPrice(NamedTuple):
+    from_asset: str
+    to_asset: str
+    exchange: str
+    historical_data: Optional[HistoricalBar] = None
+
+
+class PairConverterPlugin(AbstractPairConverterPlugin):
+    def __init__(self, historical_price_type: str) -> None:
+        super().__init__(historical_price_type=historical_price_type)
+        self.__logger: logging.Logger = create_logger(f"{self.name()}/{historical_price_type}")
+
+        self.__exchanges: Dict[str, Exchange] = {}
+        self.__exchange_markets: Dict[str, Dict[str, List[str]]] = {}
+        self.__exchange_graphs: Dict[str, Dict[str, List[str]]] = {}

```

Comment:
It would be best to have generic-typed Graph and Vertex classes here, rather than nested Dicts, but we can leave this for a future improvement.